File: src-tauri/src/agents.rs
// Agent library (port of agents.ts): read/write Claude Code subagent definition
// files (.claude/agents/*.md) at user (~/.claude) and project scope.

use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use serde_yaml::Value as Yaml;

#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentDefinition {
    pub name: String,
    pub description: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tools: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub effort: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_turns: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub color: Option<String>,
    #[serde(default)]
    pub prompt: String,
    pub scope: String, // "user" | "project"
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub project_path: Option<String>,
    #[serde(default)]
    pub path: String,
}

fn user_agents_dir() -> PathBuf {
    let home = std::env::var("HOME").unwrap_or_default();
    PathBuf::from(home).join(".claude").join("agents")
}

fn split_frontmatter(raw: &str) -> (Yaml, String) {
    if let Some(rest) = raw.strip_prefix("---\n").or_else(|| raw.strip_prefix("---\r\n")) {
        if let Some(end) = rest.find("\n---") {
            let fm = &rest[..end];
            let body = rest[end + 4..].trim_start_matches(['\r', '\n']).to_string();
            let yaml = serde_yaml::from_str::<Yaml>(fm).unwrap_or(Yaml::Null);
            return (yaml, body);
        }
    }
    (Yaml::Null, raw.to_string())
}

fn yaml_str(m: &Yaml, key: &str) -> Option<String> {
    m.get(key).and_then(|v| v.as_str()).map(|s| s.to_string())
}

fn parse_agent(path: &Path, scope: &str, project_path: Option<&str>) -> Option<AgentDefinition> {
    let raw = std::fs::read_to_string(path).ok()?;
    let (fm, body) = split_frontmatter(&raw);
    let name = yaml_str(&fm, "name").filter(|n| !n.trim().is_empty())?;
    let tools = fm.get("tools").and_then(|v| match v {
        Yaml::Sequence(seq) => Some(seq.iter().filter_map(|x| x.as_str().map(|s| s.to_string())).collect()),
        Yaml::String(s) => Some(s.split(',').map(|p| p.trim().to_string()).filter(|p| !p.is_empty()).collect()),
        _ => None,
    });
    Some(AgentDefinition {
        name,
        description: yaml_str(&fm, "description").unwrap_or_default(),
        model: yaml_str(&fm, "model"),
        tools,
        effort: yaml_str(&fm, "effort"),
        max_turns: fm.get("maxTurns").and_then(|v| v.as_u64()).map(|n| n as u32),
        color: yaml_str(&fm, "color"),
        prompt: body.trim().to_string(),
        scope: scope.to_string(),
        project_path: project_path.map(|s| s.to_string()),
        path: path.to_string_lossy().to_string(),
    })
}

fn read_dir_agents(dir: &Path, scope: &str, project_path: Option<&str>, out: &mut Vec<AgentDefinition>) {
    let Ok(entries) = std::fs::read_dir(dir) else { return };
    for entry in entries.flatten() {
        let p = entry.path();
        if p.is_dir() {
            read_dir_agents(&p, scope, project_path, out);
        } else if p.extension().map(|e| e == "md").unwrap_or(false) {
            if let Some(a) = parse_agent(&p, scope, project_path) {
                out.push(a);
            }
        }
    }
}

pub fn list_agents(project_path: Option<&str>) -> Vec<AgentDefinition> {
    let mut out = vec![];
    read_dir_agents(&user_agents_dir(), "user", None, &mut out);
    if let Some(pp) = project_path {
        let dir = Path::new(pp).join(".claude").join("agents");
        read_dir_agents(&dir, "project", Some(pp), &mut out);
    }
    out.sort_by(|a, b| a.name.to_lowercase().cmp(&b.name.to_lowercase()));
    out
}

fn file_stem(name: &str) -> String {
    let s: String = name.trim().to_lowercase().chars().map(|c| if c.is_ascii_alphanumeric() || c == '-' { c } else { '-' }).collect();
    let s = s.trim_matches('-').to_string();
    if s.is_empty() { "agent".into() } else { s }
}

fn serialize_agent(def: &AgentDefinition) -> String {
    let mut map = serde_yaml::Mapping::new();
    map.insert("name".into(), def.name.clone().into());
    map.insert("description".into(), def.description.clone().into());
    if let Some(m) = &def.model { if !m.is_empty() { map.insert("model".into(), m.clone().into()); } }
    if let Some(t) = &def.tools { if !t.is_empty() { map.insert("tools".into(), Yaml::Sequence(t.iter().map(|s| s.clone().into()).collect())); } }
    if let Some(e) = &def.effort { if !e.is_empty() { map.insert("effort".into(), e.clone().into()); } }
    if let Some(n) = def.max_turns { map.insert("maxTurns".into(), (n as u64).into()); }
    if let Some(c) = &def.color { if !c.is_empty() { map.insert("color".into(), c.clone().into()); } }
    let yaml = serde_yaml::to_string(&Yaml::Mapping(map)).unwrap_or_default();
    format!("---\n{}---\n\n{}\n", yaml, def.prompt.trim())
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveResult {
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub path: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

pub fn save_agent(def: &AgentDefinition, original_path: Option<&str>) -> SaveResult {
    if def.name.trim().is_empty() {
        return SaveResult { ok: false, path: None, error: Some("Agent name is required".into()) };
    }
    if def.description.trim().is_empty() {
        return SaveResult { ok: false, path: None, error: Some("Description is required".into()) };
    }
    let dir = if def.scope == "project" {
        match &def.project_path {
            Some(pp) => Path::new(pp).join(".claude").join("agents"),
            None => return SaveResult { ok: false, path: None, error: Some("Project path is required for a project agent".into()) },
        }
    } else {
        user_agents_dir()
    };
    let target = dir.join(format!("{}.md", file_stem(&def.name)));
    if let Err(e) = std::fs::create_dir_all(&dir) {
        return SaveResult { ok: false, path: None, error: Some(e.to_string()) };
    }
    if let Err(e) = std::fs::write(&target, serialize_agent(def)) {
        return SaveResult { ok: false, path: None, error: Some(e.to_string()) };
    }
    if let Some(orig) = original_path {
        if !orig.is_empty() && Path::new(orig) != target && Path::new(orig).exists() {
            let _ = std::fs::remove_file(orig);
        }
    }
    SaveResult { ok: true, path: Some(target.to_string_lossy().to_string()), error: None }
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OkResult {
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

pub fn delete_agent(path: &str) -> OkResult {
    match std::fs::remove_file(path) {
        Ok(_) => OkResult { ok: true, error: None },
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => OkResult { ok: true, error: None },
        Err(e) => OkResult { ok: false, error: Some(e.to_string()) },
    }
}

File: electron/src/main/agents.ts
// The agent library: Claude Code subagent definitions, which are Markdown files with YAML
// frontmatter in `~/.claude/agents` (user scope) and `<project>/.claude/agents` (project
// scope). Mirrors `src-tauri/src/agents.rs`.
//
// These are Claude Code's files, not Operator's — Operator is a visual editor over them. So
// the parser is forgiving (a file that isn't a valid agent is skipped, never fatal) and the
// writer only emits keys that have values, to avoid decorating a user's file with nulls.
import { mkdir, readFile, readdir, rm, writeFile } from 'node:fs/promises'
import { homedir } from 'node:os'
import { extname, join, resolve } from 'node:path'
import { parse as parseYaml, stringify as stringifyYaml } from 'yaml'

// The renderer's own type — including the `AgentScope` union, which a local `string` widens.
import type { AgentDefinition } from '../../../src/shared/types'
export type { AgentDefinition }

const userAgentsDir = () => join(homedir(), '.claude', 'agents')

/** Split `---\n<yaml>\n---\n<body>`. Mirrors the Rust byte-for-byte, including its tolerance
 *  for CRLF and its rule that a file without frontmatter is all body. A real YAML parser does
 *  the frontmatter itself: these files are hand-written by users, and quoting, lists and
 *  multi-line scalars are exactly what a hand-rolled splitter gets wrong. */
function splitFrontmatter(raw: string): { fm: Record<string, unknown>; body: string } {
  const rest = raw.startsWith('---\n') ? raw.slice(4) : raw.startsWith('---\r\n') ? raw.slice(5) : null
  if (rest != null) {
    const end = rest.indexOf('\n---')
    if (end >= 0) {
      let fm: Record<string, unknown> = {}
      try {
        const parsed = parseYaml(rest.slice(0, end))
        if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) fm = parsed as Record<string, unknown>
      } catch { /* malformed frontmatter → treat as absent, same as the Rust's unwrap_or(Null) */ }
      return { fm, body: rest.slice(end + 4).replace(/^[\r\n]+/, '') }
    }
  }
  return { fm: {}, body: raw }
}

const str = (v: unknown): string | undefined => (typeof v === 'string' ? v : undefined)

async function parseAgent(path: string, scope: string, projectPath?: string): Promise<AgentDefinition | null> {
  let raw: string
  try { raw = await readFile(path, 'utf8') } catch { return null }
  const { fm, body } = splitFrontmatter(raw)
  const name = str(fm.name)?.trim()
  // No name, no agent. This is the check that keeps a stray README.md in the folder from
  // showing up in the library as a nameless row.
  if (!name) return null

  // `tools:` is a list in some files and a comma-separated string in others; Claude Code
  // accepts both, so both are read.
  let tools: string[] | undefined
  if (Array.isArray(fm.tools)) tools = fm.tools.filter((t): t is string => typeof t === 'string')
  else if (typeof fm.tools === 'string') tools = fm.tools.split(',').map((p) => p.trim()).filter(Boolean)

  return {
    name,
    description: str(fm.description) ?? '',
    model: str(fm.model),
    tools,
    effort: str(fm.effort),
    maxTurns: typeof fm.maxTurns === 'number' ? fm.maxTurns : undefined,
    color: str(fm.color),
    prompt: body.trim(),
    scope: scope as AgentDefinition['scope'],
    projectPath,
    path,
  }
}

async function readDirAgents(dir: string, scope: string, projectPath: string | undefined, out: AgentDefinition[]): Promise<void> {
  let entries
  try { entries = await readdir(dir, { withFileTypes: true }) } catch { return }
  for (const e of entries) {
    const p = join(dir, e.name)
    if (e.isDirectory()) await readDirAgents(p, scope, projectPath, out)
    else if (extname(e.name) === '.md') {
      const a = await parseAgent(p, scope, projectPath)
      if (a) out.push(a)
    }
  }
}

export async function listAgents(projectPath?: string): Promise<AgentDefinition[]> {
  const out: AgentDefinition[] = []
  await readDirAgents(userAgentsDir(), 'user', undefined, out)
  if (projectPath) await readDirAgents(join(projectPath, '.claude', 'agents'), 'project', projectPath, out)
  return out.sort((a, b) => a.name.toLowerCase().localeCompare(b.name.toLowerCase()))
}

/** Filename from the agent's name: lowercase, non-alphanumerics to `-`, trimmed. Also the
 *  reason `save` takes an `originalPath` — renaming an agent changes its filename, so the old
 *  file has to be removed or the library shows both. */
function fileStem(name: string): string {
  const s = name.trim().toLowerCase().replace(/[^a-z0-9-]/g, '-').replace(/^-+|-+$/g, '')
  return s || 'agent'
}

function serializeAgent(def: AgentDefinition): string {
  const fm: Record<string, unknown> = { name: def.name, description: def.description }
  if (def.model) fm.model = def.model
  if (def.tools?.length) fm.tools = def.tools
  if (def.effort) fm.effort = def.effort
  if (def.maxTurns != null) fm.maxTurns = def.maxTurns
  if (def.color) fm.color = def.color
  return `---\n${stringifyYaml(fm)}---\n\n${def.prompt.trim()}\n`
}

export async function saveAgent(def: AgentDefinition, originalPath?: string): Promise<{ ok: boolean; path?: string; error?: string }> {
  if (!def.name?.trim()) return { ok: false, error: 'Agent name is required' }
  if (!def.description?.trim()) return { ok: false, error: 'Description is required' }
  let dir: string
  if (def.scope === 'project') {
    if (!def.projectPath) return { ok: false, error: 'Project path is required for a project agent' }
    dir = join(def.projectPath, '.claude', 'agents')
  } else {
    dir = userAgentsDir()
  }
  const target = join(dir, `${fileStem(def.name)}.md`)
  try {
    await mkdir(dir, { recursive: true })
    await writeFile(target, serializeAgent(def), 'utf8')
  } catch (e) {
    return { ok: false, error: String(e) }
  }
  // A rename leaves the old file behind; remove it only once the new one is safely written,
  // so a failed write never costs the user the original.
  if (originalPath && resolve(originalPath) !== resolve(target)) {
    await rm(originalPath, { force: true }).catch(() => { /* already gone */ })
  }
  return { ok: true, path: target }
}

/** A missing file is SUCCESS, not an error: the caller wanted it gone and it is gone. */
export async function deleteAgent(path: string): Promise<{ ok: boolean; error?: string }> {
  try {
    await rm(path, { force: true })
    return { ok: true }
  } catch (e) {
    return { ok: false, error: String(e) }
  }
}
